GAD 9 - AVL Baum/tests/avl/AVLTreeTester.java

```diff
@@ -0,0 +1,316 @@
+
+        b.setLeft(new AVLTreeNode(-3));
+        b.setRight(b2);
+        assertFalse(tree.validAVL(), "all nodes of all subtrees need to be sorted");
+
+        b.setLeft(b1);
+        b.setRight(null);
+        assertFalse(tree.validAVL(), "wrong balance at node B");
+
+        b.setBalance(-1);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+
+        root.setBalance(0);
+        a.setBalance(1);
+        assertFalse(tree.validAVL(), "wrong balance at node A");
+
+        a.setLeft(null);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+    }
+
+    /**
+     * Count the number of nodes in the subtree, including the root itself.
+     *
+     * @param subtree root of a subtree
+     * @return number of nodes in the subtree
+     */
+    private int count(AVLTreeNode subtree) {
+        int l = subtree.getLeft() == null ? 0 : count(subtree.getLeft());
+        int r = subtree.getRight() == null ? 0 : count(subtree.getRight());
+        return 1 + l + r;
+    }
+
+    /**
+     * Test to check that insert() works to insert new values.
+     * No checks on the layout of the AVL tree are performed in
+     * this test. It will only check that the number of nodes
+     * in the tree is always correct after every insert() call.
+     */
+    @Test
+    public void testInsertNumberOfValues() {
+        AVLTree tree = new AVLTree();
+        tree.insert(42);
+        try {
+            assertEquals(42, tree.getRoot().getKey(), "root node has wrong value");
+        } catch (NullPointerException exc) {
+            fail("you need to implement insert() first");
+        }
+
+        assertEquals(1, count(tree.getRoot()), "expected one element in the tree");
+        Random random = new Random();
+        for (int i = 0; i < 999; i++) {
+            tree.insert(random.nextInt());
+            assertEquals(2 + i, count(tree.getRoot()), "expected " + (2 + i) + " nodes in the tree");
+        }
+    }
+
+    /**
+     * Test to check whether the combination of find() and insert()
+     * work as expected. Note that no tree layout checks are performed
+     * by this unit test, it will actually only verify the find()
+     * method. But that's enough to make the test fail if insert() does
+     * not work. Also note that multiple same keys are not checked.
+     */
+    @Test
+    public void testFindInsert() {
+        AVLTree tree = new AVLTree();
+        AVLTreeNode root = new AVLTreeNode(12);
+        tree.setRoot(root);
+        assertTrue(tree.find(12));
+        assertFalse(tree.find(5));
+        assertFalse(tree.find(0));
+        assertFalse(tree.find(-12));
+        tree.setRoot(null);
+
+        Random random = new Random();
+        List<Integer> values = new ArrayList<>();
+        for (int i = 0; i < 1000; i++) {
+            int v = random.nextInt(2000) - 1000;
+            values.add(v);
+            tree.insert(v);
+            assertTrue(tree.find(v));
+        }
+
+        for (int i = -1000; i < 1000; i++) {
```

Comment:
I would suggest to loop first through all `values` and after that generate new random numbers, and test if those can't be found when they are not in `values`.
An explanatory message would also be nice here.

Reply:
I've added an explanatory message and increased the bounds of the checked values to ensure that some are definitively not in the tree nor in the list `values`. But I didn't add the extra loop, since the second one already goes through the whole list of values.

Reply:
Yeah, but this does not ensure that during insertion, some values aren't deleted.

Reply:
Which values could be deleted upon insertion? I would not expect any value to be deleted tbh
How could you check that this doesn't happen? Create a list of all `AVLTreeNode` objects and all those entries must be present after inserting one more?

Reply:
I do not know how values could be deleted. But If we are testing wether some values are or are not in the tree, it would make sense to at least test for all values that should be in the tree if we have a list of them.

Reply:
There's already such a check using the `values` list, if I understood you correctly. But we can't really check for the `AVLTreeNode` objects, since they are hidden behind the `insert` method of the tree. I've added a check to ensure that there are exactly 1000 nodes in the tree, though. This would somewhat address your issue, even though it does not completely ensure that some implementation doesn't replace nodes.